### PR TITLE
fix(messages): report Anthropic streaming usage from the trailing usage-only chunk

### PR DESCRIPTION
**Pull request by Syrunekai**
## Description

When bridging an OpenAI-compatible provider to the Anthropic Messages API, streaming via `amessages` reported zero token usage. OpenAI-compatible providers emit token counts in a final usage-only chunk that arrives *after* the `finish_reason` chunk, but the bridge emitted the closing `message_delta`/`message_stop` on the `finish_reason` chunk — before that usage was available — so `message_delta` always reported `input_tokens=0`/`output_tokens=0` (and no cache).

The fix defers the closing `message_delta`/`message_stop` to the stream wrapper's post-loop flush so usage is complete by the time they're emitted, and maps `prompt_tokens_details.cached_tokens` → `cache_read_input_tokens`. Since the converter no longer emits those closing events, it also removes the now-dead `emitted_stop` bookkeeping in the wrapper and narrows the converter's return type to the events it actually produces.

**Verified** against multiple OpenAI-compatible providers (including OpenRouter and Anthropic's OpenAI-compatible endpoint), and cross-checked against the native Anthropic provider as a reference for usage/cache placement — native Anthropic reports cache in `message_start`, while the bridge necessarily reports it in `message_delta` because OpenAI-compatible usage arrives last (both are valid per the SDK). New unit tests added; full unit suite + `pre-commit` (ruff + mypy strict) pass locally.

## PR Type

- 🐛 Bug Fix

## Relevant issues

None found.

## Checklist

- [x] I understand the code I am submitting.
- [x] I have added unit tests that prove my fix/feature works
- [x] I have run this code locally and verified it fixes the issue.
- [x] New and existing tests pass locally
- [ ] Documentation was updated where necessary 
- [x] I have read and followed the [contribution guidelines](https://github.com/mozilla-ai/any-llm/blob/main/CONTRIBUTING.md)
- [x] **AI Usage:**
    - [ ] No AI was used.
    - [ ] AI was used for drafting/refactoring.
    - [x] This is fully AI-generated.

## AI Usage Information

- AI Model used: Claude Opus 4.8
- AI Developer Tool used: Claude Code
- Any other info you'd like to share: AI-authored, but human-directed and reviewed change-by-change. On your "discuss with the human, not the AI" policy — fully respected: the strings here are pulled by a human. For any review discussion that needs a real conversation, my puppeteer @0xSylice will reply personally (not paste my answers back).

- [x] I am an AI Agent filling out this form (check box if true)

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved streaming event ordering so final stop reasons are emitted reliably, including when a stream ends with trailing usage-only chunks.
  * Streaming usage is now flushed correctly on normal completion and mid-stream failures, and cached input tokens are reported when available.
  * Standardised stop-reason handling across streaming conversions, with consistent defaults when no finish reason is provided.
* **Tests**
  * Expanded unit and compatibility tests for cached-token scenarios (present/zero/missing), stop-reason semantics, and robustness during exceptions and early closure.

### CODE REVIEW

**Comment by Syrunekai** [comment]
Thanks @coderabbitai — checked each against the repo's config and conventions:

**1. Split the converter (PLR0912):** the repo's `[tool.ruff.lint]` config ignores `PLR`/`PLR0912`, so the project's own linter doesn't flag this and pre-commit is green. I'd prefer not to restructure the converter inside a focused bug-fix PR — glad to do it as a separate change if the maintainers want it.

**2. `is not None` for usage values:** the `prompt_tokens`/`completion_tokens` checks are pre-existing and outside this change. For the `cached_tokens` line I added, truthy vs `is not None` produces the same result for the single-usage-chunk case providers actually emit (and the wrapper does `state.cache_read_input_tokens or None`, so a `0` collapses to an omitted field either way); the zero case is covered by a new unit test. Left as-is to keep the diff surgical — happy to switch if you'd prefer the explicit `None` check.

**3. Imports to top / `-> Any` (ANN401):** the repo's ruff config ignores `ANN`, and this test matches the file's established pattern — inline imports throughout, and an identical `mock_stream() -> Any` in the sibling test (`test_default_amessages_streaming`, ~line 356). Kept it consistent with the file.

Happy to make any of these if a maintainer sees them as blocking.

**Comment by 0xSylice** [comment]
Human in the loop, reporting in.

This is a fairly simple fix. When a response is streamed out through the Anthropic messages format, the token usage numbers (input, output, and how many were cached) only land right at the very end of the stream.

The code was closing out and sending the usage report a bit too early, before those final numbers had actually arrived, so it always returned as zero.

The fix just waits until the stream is completely finished before reporting, so the real values (cached count included) are actually there. That's pretty much it!

- Sylice

File: src/any_llm/any_llm.py [hunk 1]
```diff
@@ -785,27 +785,23 @@ async def _amessages(
...
                     type="message_delta",
-                    delta=MessageDelta(stop_reason="end_turn"),
+                    delta=MessageDelta(stop_reason=state.stop_reason or "end_turn"),
                     usage=MessageDeltaUsage(
                         output_tokens=state.output_tokens,
                         input_tokens=state.input_tokens,
+                        cache_read_input_tokens=state.cache_read_input_tokens or None,
                     ),
                 )
                 yield MessageStopEvent(type="message_stop")
```
Comment (tbille) [inline src/any_llm/any_llm.py]: **Usage is lost when the stream fails mid-iteration.** These closing events are the only ones carrying usage, and they are now reached only if `async for` completes normally. If `result` raises during iteration, this block is skipped and the caller gets zero usage despite `state` already holding accumulated tokens.

Wrap the loop so the closing events are flushed on failure too, while still handling early consumer-close (`GeneratorExit`, where yielding is illegal) and avoiding double emission:

```python
async def convert_stream() -> AsyncIterator[MessageStreamEvent]:
    state = StreamingState()
    emitted_closing = False

    def closing_events() -> list[MessageStreamEvent]:
        events: list[MessageStreamEvent] = []
        if not state.started:
            return events
        if state.current_block_type is not None:
            events.append(ContentBlockStopEvent(type="content_block_stop", index=state.current_block_index))
        events.append(
            MessageDeltaEvent(
                type="message_delta",
                delta=MessageDelta(stop_reason=state.stop_reason or "end_turn"),
                usage=MessageDeltaUsage(
                    output_tokens=state.output_tokens,
                    input_tokens=state.input_tokens,
                    cache_read_input_tokens=state.cache_read_input_tokens or None,
                ),
            )
        )
        events.append(MessageStopEvent(type="message_stop"))
        return events

    try:
        async for chunk in result:
            for event in chat_completion_chunk_to_message_stream_events(chunk, state):
                yield event
        emitted_closing = True
        for event in closing_events():
            yield event
    except GeneratorExit:
        # Consumer abandoned the stream; yielding here is illegal and no one is left to receive usage.
        raise
    except BaseException:
        # Provider stream failed mid-iteration: flush usage so the caller still sees tokens consumed, then re-raise.
        if not emitted_closing:
            for event in closing_events():
                yield event
        raise
```

Two things worth a team decision:
1. Whether the failure path should emit `message_stop` at all, since Anthropic semantics reserve it for clean completion. If not, emit only the usage-bearing `message_delta` on the error path.
2. On early failure, usage may be partial/zero, but that is still strictly better than emitting nothing.

File: tests/unit/test_messages.py [hunk 2]
```diff
@@ -396,6 +396,70 @@ async def mock_stream() -> Any:
...
+    assert not isinstance(result, (MessageResponse, ParsedMessage))
+
+    events = [event async for event in result]
+    delta = next(e for e in events if isinstance(e, MessageDeltaEvent))
+    assert delta.usage.input_tokens == 100
+    assert delta.usage.output_tokens == 50
+    assert delta.usage.cache_read_input_tokens == 80
+    assert delta.delta.stop_reason == "end_turn"
+    assert events[-1].type == "message_stop"
+
```
Comment (tbille) [inline tests/unit/test_messages.py:460]: The new tests only cover the success path. Please add coverage for the failure path that this PR's requirement depends on:

1. A stream that yields some content chunks (populating `state` usage) and then **raises** mid-iteration should still emit a `message_delta` carrying the accumulated usage before the exception propagates.
2. A consumer that closes/abandons the generator early (triggering `GeneratorExit`) must not raise `RuntimeError` from yielding inside the cleanup path.

Sketch for (1):

```python
async def mock_stream() -> Any:
    yield ChatCompletionChunk(
        id="c1", model="gpt-4", created=0, object="chat.completion.chunk",
        choices=[ChunkChoice(index=0, delta=ChoiceDelta(content="Hi"), finish_reason=None)],
        usage=CompletionUsage(prompt_tokens=100, completion_tokens=10, total_tokens=110),
    )
    raise RuntimeError("provider stream dropped")

result = await AnyLLM._amessages(mock_provider, params)
seen: list[Any] = []
with pytest.raises(RuntimeError):
    async for event in result:
        seen.append(event)
delta = next(e for e in seen if isinstance(e, MessageDeltaEvent))
assert delta.usage.input_tokens == 100
assert delta.usage.output_tokens == 10
```

**Review by tbille — CHANGES_REQUESTED** [review]
Thanks for this fix. The core change is correct for its stated goal: OpenAI-compatible providers emit usage in a trailing usage-only chunk that arrives after `finish_reason`, and deferring the closing `message_delta`/`message_stop` to the post-loop flush is the right way to capture it. The `cache_read_input_tokens` mapping and the `state.stop_reason` plumbing are good, and the added tests cover the happy path plus the cache edge cases well.

There is one regression I'd like addressed before merge: moving the usage-bearing closing events to *after* the `async for` loop means they are only emitted on a clean completion. If the underlying provider stream raises mid-iteration (network drop, provider 5xx, timeout, or the trailing usage chunk itself failing to arrive), the exception propagates out of the loop and the `if state.started:` block is skipped, so the consumer receives **no** `message_delta` and loses the tokens already accumulated in `state`. This is amplified by `@handle_exceptions(wrap_streaming=True)`, whose `_wrap_async_iterator` re-raises on iteration errors.

Before this PR, usage was emitted inline on the `finish_reason` chunk, so a failure *after* finish_reason still left the caller with a usage-bearing `message_delta`. The PR trades that resilience for trailing-chunk correctness without preserving the failure path. We should guarantee tokens are reported even when the stream fails. Inline suggestions below.

**Comment by Syrunekai** [comment]
Thanks @tbille — confirmed the regression exactly as you traced it: with the closing events moved after the loop, a mid-iteration raise skipped the `if state.started:` block entirely, and `_wrap_async_iterator` re-raises, so the caller lost everything `state` had accumulated. Agreed it needed addressing before merge.

Pushed d769735 implementing the structure you suggested (deferred closing events, flushed on failure too), with three deliberate deviations from the sketch — each is a one-line change to revert if the team prefers the original:

1. **`except Exception` instead of the `GeneratorExit` guard + `except BaseException`.** `GeneratorExit` and `asyncio.CancelledError` are `BaseException` subclasses, so they propagate untouched: the yield-during-cleanup hazard can't occur by construction, and cancellation isn't delayed by yielding events into a consumer that's being torn down. It also matches the `Exception`-only catch in `handle_exceptions._wrap_async_iterator`, which wraps this same stream one level up.

2. **The failure-path `stop_reason` is reported as-known rather than defaulted to `"end_turn"`.** When the failure hits after `finish_reason` arrived (the resilience case you called out, e.g. the trailing usage chunk request dying), `state.stop_reason` is set and the real value is reported, same as the sketch. When the stream dies before `finish_reason`, defaulting to `"end_turn"` would present a mid-generation failure as a natural completion; the SDK types `Delta.stop_reason` as `Optional`, and `None` is the accurate value there. The `or "end_turn"` default remains on the success path, where it covers providers that end cleanly without a `finish_reason` (pre-existing behavior).

3. **On your question 1 — the failure path emits only the usage-bearing `message_delta`: no `message_stop`, and no `content_block_stop` either.** Agree that `message_stop` belongs to clean completion, and there's a concrete failure mode behind it: a consumer that idiomatically stops iterating at `message_stop` closes the generator at that yield, so the pending re-raise never executes — the provider error is silently swallowed and a truncated message is presented as complete. With `message_stop` omitted, the same consumer keeps iterating and receives the exception on the next `__anext__`. `content_block_stop` is dropped by the same reasoning one level down: it asserts a block completed when it was actually severed mid-generation, and unlike `message_delta` it carries no data a consumer would otherwise lose. The rule the implementation follows: on failure, emit exactly the event that carries otherwise-lost data (the usage delta) and nothing whose only function is to assert a completion that didn't happen. For reference, the native Anthropic stream doesn't close open blocks on failure either — it emits an `error` event and ends — so consumers can't be relying on balanced events under failure.

Tests: both scenarios from your review are covered — the mid-iteration raise asserts the flushed `message_delta` carries the accumulated usage (and that no `message_stop`/`content_block_stop` follows) before the exception propagates, and early consumer close is asserted clean at both suspension points: during normal iteration and at the failure-flush yield inside the `except` block. Also added: failure before the first chunk emits nothing (no events before `message_start`), and two small tests for pre-existing uncovered branches that this diff's hunks re-emit (`content_block_stop` when a stream ends with a block open, and the empty-stream case) — Codecov's patch gate reads hunks rather than semantic changes, so this keeps the check green instead of flagging lines the fix didn't behaviorally alter.

All three calls above are your team's to make — each variant is a one-line change to the closing-events helper, happy to switch.

**Review by dpoulopoulos — DISMISSED** [review]

**Review by tbille — DISMISSED** [review]
Thanks for the changes @Syrunekai 
Apologies for the delay, merging it and will trigger a release.

File: src/any_llm/any_llm.py [hunk 3]
```diff
@@ -785,29 +786,35 @@ async def _amessages(
...
+
+            try:
+                async for chunk in result:
+                    for event in chat_completion_chunk_to_message_stream_events(chunk, state):
+                        yield event
+            except Exception:
+                # Flush the usage accumulated so far before re-raising, so a mid-stream failure still reports tokens.
+                if state.started:
+                    yield usage_delta(state.stop_reason)
+                raise
```
[the remark this answers is not in the document]
Comment (tbille) [reply src/any_llm/any_llm.py:811]: Valid finding, fixed in 4757575. The failure path now flushes usage with `stop_reason=None` unconditionally (via `usage_delta(None)`), so a `finish_reason` seen before the failure can no longer be mistaken for a successful completion. `message_stop` remains reserved for the clean-completion path.

Added regression test `test_default_amessages_streaming_failure_after_finish_reason_reports_no_stop_reason`: the stream emits `finish_reason="stop"` then raises before the trailing usage chunk, and asserts the flushed `message_delta` reports `stop_reason is None` (with usage still preserved) and no `message_stop` is emitted.

Note: `content_block_stop` does legitimately appear in this scenario because the content block is closed when the `finish_reason` chunk is processed during normal iteration, so the test asserts only on the absence of `message_stop`, not `content_block_stop`.

**Review by tbille — APPROVED** [review]
